addons/gamepad/userscript.js

```diff
@@ -71,27 +70,66 @@ export default async function ({ addon, console, msg }) {
       return null;
     }
     const jsonText = lineWithMagic.substr(0, lineWithMagic.length - GAMEPAD_CONFIG_MAGIC.length);
-    let parsed;
+    let storedSettingsMap = null;
     try {
-      parsed = JSON.parse(jsonText);
-      if (!parsed || typeof parsed !== "object" || !Array.isArray(parsed.buttons) || !Array.isArray(parsed.axes)) {
+      const parsed = JSON.parse(jsonText);
+      if (!parsed || typeof parsed !== "object") {
+        throw new Error("Invalid data");
+      }
+
+      // The older version of this addon only saved one set of settings with 2 arrays: .buttons and .axes
+      // This version saves an array of multiple gamepads and their settings
+      //   This code determines whether the project has an older comment (V1) in it or a newer comment (V2)
+      if (Array.isArray(parsed)) {
+        // Comment is in the new V2 style
+        // Convert the array back into a Map
+        storedSettingsMap = new Map(parsed);
+      } else if (Array.isArray(parsed.buttons) && Array.isArray(parsed.axes)) {
+        // Comment is in the old V1 style
+        // Convert to V2 style as a Map containing a single unlabeled gamepad
+        storedSettingsMap = new Map([
+          [
+            "", // No gamepad type/key to store
+            {
+              id: "", // No gamepad id to store
+              index: 0,
+              settings: {
+                axes: parsed.axes,
+                buttons: parsed.buttons,
+              },
+            },
+          ],
+        ]);
+      } else {
         throw new Error("Invalid data");
       }
     } catch (e) {
       console.warn("Gamepad comment has invalid JSON", e);
       return null;
     }
-    return parsed;
+    return storedSettingsMap;
   };
 
   GamepadLib.setConsole(console);
   const gamepad = new GamepadLib();
 
-  gamepad.getUserHints = () => {
-    const parsedOptions = parseOptionsComment();
-    if (parsedOptions) {
+  gamepad.getUserHints = (gamepadId, gamepadIndex) => {
+    const parsedOptionsMap = parseOptionsComment(findOptionsComment());
+    if (parsedOptionsMap) {
+      const optionsArray = Array.from(parsedOptionsMap.values());
+      let matchingOption;
+      // If a stored mapping exists for this exact gamepad id and index, use it
+      matchingOption = optionsArray.find((gamepads) => gamepads.id === gamepadId && gamepads.index === gamepadIndex);
+      if (!matchingOption) {
+        // Otherwise, if a stored mapping exists for this gamepad id at all, use it
+        matchingOption = optionsArray.find((gamepads) => gamepads.id === gamepadId);
+        if (!matchingOption) {
+          // Otherwise, use the first stored mapping no matter what it is
+          matchingOption = optionsArray.at(0);
+        }
+      }
```

Comment:
I would expect that the configuration of the same index be preferred over the first one even if the gamepad isn't the same. Oftentimes, a player's gamepad won't have the exact same ID as the project creator's gamepad, and it can even vary between browsers.

```suggestion
      // If a stored mapping exists for this exact gamepad id and index, use it
      matchingOption = optionsArray.find((gamepads) => gamepads.id === gamepadId && gamepads.index === gamepadIndex);
      if (!matchingOption) {
        // Otherwise, if a stored mapping exists for this gamepad id at all, use it
        matchingOption = optionsArray.find((gamepads) => gamepads.id === gamepadId);
        if (!matchingOption) {
          // Otherwise, use the stored mapping at the same index if it exists
          matchingOption = optionsArray.find((gamepads) => gamepads.index === gamepadIndex);
          if (!matchingOption) {
            // Otherwise, use the first stored mapping no matter what it is
            matchingOption = optionsArray.at(0);
          }
        }
      }
```

Reply:
> I would expect that the configuration of the same index be preferred over the first one even if the gamepad isn't the same. Oftentimes, a player's gamepad won't have the exact same ID as the project creator's gamepad, and it can even vary between browsers.

Good point, I didn't think about that. I just committed your changes